driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -656,82 +640,14 @@ public void onResult(final ByteBuf result, final Throwable t) {
-                    commandEventException = new MongoCommandException(new BsonDocument(), description.getServerAddress());
-                }
-                sendCommandFailedEvent(message, lazyCommandDocument.getName(), description, startTimeNanos, commandEventException,
-                        commandListener);
-            }
-        }
-
-        public void sendSucceededEvent(final ResponseBuffers responseBuffers) {
-            if (commandListener != null && opened()) {
-                BsonDocument responseDocumentForEvent = (SECURITY_SENSITIVE_COMMANDS.contains(lazyCommandDocument.getName()))
-                        ? new BsonDocument()
-                        : getResponseDocument(responseBuffers, message.getId(),
-                        new RawBsonDocumentCodec());
-                sendCommandSucceededEvent(message, lazyCommandDocument.getName(), responseDocumentForEvent, description,
-                        startTimeNanos, commandListener);
-            }
-        }
-
-        public void sendSucceededEventForOneWayCommand() {
-            if (commandListener != null && opened()) {
-                BsonDocument responseDocumentForEvent = new BsonDocument("ok", new BsonInt32(1));
-                sendCommandSucceededEvent(message, lazyCommandDocument.getName(), responseDocumentForEvent, description,
-                        startTimeNanos, commandListener);
-            }
-        }
-    }
-
-    // Lazily determine the command document and command name, since they're only needed if either a command listener or compression
-    // is enabled
-    private static final class LazyCommandDocument {
-        private final CommandMessage commandMessage;
-        private final ByteBufferBsonOutput bsonOutput;
-        private BsonDocument commandDocument;
-        private String commandName;
-
-        private LazyCommandDocument(final CommandMessage commandMessage, final ByteBufferBsonOutput bsonOutput) {
-            this.commandMessage = commandMessage;
-            this.bsonOutput = bsonOutput;
-        }
-
-        public String getName() {
-            if (commandName == null) {
-                commandName = getDocument().getFirstKey();
-            }
-            return commandName;
-        }
+    private static final Logger COMMAND_PROTOCOL_LOGGER = Loggers.getLogger("protocol.command");
 
-        private BsonDocument getDocument() {
-            if (commandDocument == null) {
-                commandDocument = commandMessage.getCommandDocument(bsonOutput);
-            }
-            return commandDocument;
+    private CommandEventSender createCommandEventSender(final CommandMessage message, final ByteBufferBsonOutput bsonOutput) {
+        if (opened() && (commandListener != null || COMMAND_PROTOCOL_LOGGER.isDebugEnabled())) {
+            return new LoggingCommandEventSender(SECURITY_SENSITIVE_COMMANDS, description, commandListener, message, bsonOutput,
```

Comment:
Separating into two implementations of an interface simplifies the LoggingCommandEventSender

Reply:
The behavioural change here is we no longer output the debug logging if a user has their own listener and debug level logging is on.

Reply:
I think we do, as it's an OR not an AND.

Reply:
You're correct! I misread.